File: RobotCode/snobot2017/src/com/snobot2017/drivetrain/ASnobotDrivetrain.java
package com.snobot2017.drivetrain;

import com.snobot.lib.logging.ILogger;
import com.snobot2017.SmartDashBoardNames;
import com.snobot2017.SnobotActor.ISnobotActor;
import com.snobot2017.joystick.IDriverJoystick;

import edu.wpi.first.wpilibj.RobotDrive;
import edu.wpi.first.wpilibj.SpeedController;
import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;

/**
 * Base class for drivetrains.  Handles most of the interaction with drivetrain
 * @author PJ
 *
 * @param <SpeedControllerType>
 */
public abstract class ASnobotDrivetrain<SpeedControllerType extends SpeedController> implements IDriveTrain
{
    protected final SpeedControllerType mLeftMotor;
    protected final SpeedControllerType mRightMotor;
    protected final IDriverJoystick mDriverJoystick;

    protected final ILogger mLogger;

    protected final RobotDrive mRobotDrive;

    protected double mLeftMotorSpeed;
    protected double mRightMotorSpeed;

    protected double mRightMotorDistance;
    protected double mLeftMotorDistance;
    
    protected int mRightEncoderRaw;
    protected int mLeftEncoderRaw;
    
    private ISnobotActor mSnobotActor;

    public ASnobotDrivetrain(
            SpeedControllerType aLeftMotor, SpeedControllerType aRightMotor,
            IDriverJoystick aDriverJoystick, 
            ILogger aLogger)
    {
        mLeftMotor = aLeftMotor;
        mRightMotor = aRightMotor;
        mRobotDrive = new RobotDrive(mLeftMotor, mRightMotor);
        
        mDriverJoystick = aDriverJoystick;
        mLogger = aLogger;

        mRobotDrive.setSafetyEnabled(false);
    }

    @Override
    public void initializeLogHeaders()
    {
        mLogger.addHeader("LeftEncoderDistance");
        mLogger.addHeader("RightEncoderDistance");
        mLogger.addHeader("LeftMotorSpeed");
        mLogger.addHeader("RightMotorSpeed");
    }

    @Override
    public void control()
    {
        if ((mSnobotActor == null) || !mSnobotActor.isInAction())
        {
            setLeftRightSpeed(mDriverJoystick.getLeftSpeed(), mDriverJoystick.getRightSpeed());
        }
    }

    @Override
    public void updateSmartDashboard()
    {
        SmartDashboard.putNumber(SmartDashBoardNames.sLEFT_DRIVE_ENCODER_RAW, mLeftEncoderRaw);
        SmartDashboard.putNumber(SmartDashBoardNames.sRIGHT_DRIVE_ENCODER_RAW, mRightEncoderRaw);
        SmartDashboard.putNumber(SmartDashBoardNames.sRIGHT_DRIVE_ENCODER_DISTANCE, mRightMotorDistance);
        SmartDashboard.putNumber(SmartDashBoardNames.sLEFT_DRIVE_ENCODER_DISTANCE, mLeftMotorDistance);
        SmartDashboard.putNumber(SmartDashBoardNames.sLEFT_DRIVE_MOTOR_SPEED, mLeftMotorSpeed);
        SmartDashboard.putNumber(SmartDashBoardNames.sRIGHT_DRIVE_MOTOR_SPEED, mRightMotorSpeed);
    }

    @Override
    public void updateLog()
    {
        mLogger.updateLogger(mLeftMotorDistance);
        mLogger.updateLogger(mRightMotorDistance);
        mLogger.updateLogger(mLeftMotorSpeed);
        mLogger.updateLogger(mRightMotorSpeed);
    }

    @Override
    public double getRightDistance()
    {
        return mRightMotorDistance;
    }

    @Override
    public double getLeftDistance()
    {
        return mLeftMotorDistance;
    }

    @Override
    public void setLeftRightSpeed(double aLeftSpeed, double aRightSpeed)
    {
        mLeftMotorSpeed = aLeftSpeed;
        mRightMotorSpeed = aRightSpeed;
        mRobotDrive.setLeftRightMotorOutputs(mLeftMotorSpeed, mRightMotorSpeed);
    }

    @Override
    public double getLeftMotorSpeed()
    {
        return mLeftMotorSpeed;
    }
    
    @Override
    public double getRightMotorSpeed()
    {
        return mRightMotorSpeed;
        
    }

    @Override
    public void stop()
    {
        setLeftRightSpeed(0, 0);
    }

    @Override
    public void setSnobotActor(ISnobotActor aSnobotActor)
    {
        mSnobotActor = aSnobotActor;

    }

}

File: Simulator/TestCpp/src/Drivetrain/ASnobotDrivetrain.cpp
/*
 * ASnobotDrivetrain.cpp
 *
 *  Created on: May 19, 2017
 *      Author: preiniger
 */

#include "Drivetrain/ASnobotDrivetrain.h"
#include "SmartDashboardNames.h"

// WpiLib
#include "SmartDashboard/SmartDashboard.h"

ASnobotDrivetrain::ASnobotDrivetrain(const std::shared_ptr<SpeedController>& aLeftMotor, const std::shared_ptr<SpeedController>& aRightMotor, const std::shared_ptr<IDriverJoystick>& aJoystick) :
<<<<<<< HEAD
<<<<<<< HEAD
        mLeftMotor(aLeftMotor),
        mRightMotor(aRightMotor), mDriverJoystick(aJoystick), mRobotDrive(aLeftMotor, aRightMotor),
        mLeftMotorSpeed(0), mRightMotorSpeed(0), mRightMotorDistance(0), mLeftMotorDistance(0), mRightEncoderRaw(0), mLeftEncoderRaw(0)
=======
        mLeftMotor(aLeftMotor), mRightMotor(aRightMotor), mDriverJoystick(aJoystick), mLeftMotorSpeed(0), mRightMotorSpeed(0), mRightMotorDistance(0), mLeftMotorDistance(0), mRightEncoderRaw(0), mLeftEncoderRaw(
                0), mRobotDrive(aLeftMotor, aRightMotor)
>>>>>>> 9fafcd1... Adding a port of the robot code in CPP
=======
        mLeftMotor(aLeftMotor),
<<<<<<< HEAD
		mRightMotor(aRightMotor),
		mDriverJoystick(aJoystick),
		mRobotDrive(aLeftMotor, aRightMotor),

		mLeftMotorSpeed(0),
		mRightMotorSpeed(0),
		mRightMotorDistance(0),
		mLeftMotorDistance(0),
		mRightEncoderRaw(0),
		mLeftEncoderRaw(0)
>>>>>>> 9bd993d... Linux Simulator
=======
        mRightMotor(aRightMotor), mDriverJoystick(aJoystick), mRobotDrive(aLeftMotor, aRightMotor),
        mLeftMotorSpeed(0), mRightMotorSpeed(0), mRightMotorDistance(0), mLeftMotorDistance(0), mRightEncoderRaw(0), mLeftEncoderRaw(0)
>>>>>>> 408e130... Adding more CAN simulation
{

}

ASnobotDrivetrain::~ASnobotDrivetrain()
{

}


void ASnobotDrivetrain::initializeLogHeaders()
{
<<<<<<< HEAD
<<<<<<< HEAD
=======
>>>>>>> 408e130... Adding more CAN simulation
    mLogger->addHeader("LeftEncoderDistance");
    mLogger->addHeader("RightEncoderDistance");
    mLogger->addHeader("LeftMotorSpeed");
    mLogger->addHeader("RightMotorSpeed");
<<<<<<< HEAD
=======
//    mLogger.addHeader("LeftEncoderDistance");
//    mLogger.addHeader("RightEncoderDistance");
//    mLogger.addHeader("LeftMotorSpeed");
//    mLogger.addHeader("RightMotorSpeed");
>>>>>>> 9fafcd1... Adding a port of the robot code in CPP
=======
>>>>>>> 408e130... Adding more CAN simulation
}

void ASnobotDrivetrain::control()
{
    setLeftRightSpeed(mDriverJoystick->getLeftSpeed(), mDriverJoystick->getRightSpeed());
}

void ASnobotDrivetrain::updateSmartDashboard()
{
    SmartDashboard::PutNumber(SmartDashboardNames::sLEFT_DRIVE_ENCODER_RAW, mLeftEncoderRaw);
    SmartDashboard::PutNumber(SmartDashboardNames::sRIGHT_DRIVE_ENCODER_RAW, mRightEncoderRaw);
    SmartDashboard::PutNumber(SmartDashboardNames::sRIGHT_DRIVE_ENCODER_DISTANCE, mRightMotorDistance);
    SmartDashboard::PutNumber(SmartDashboardNames::sLEFT_DRIVE_ENCODER_DISTANCE, mLeftMotorDistance);
    SmartDashboard::PutNumber(SmartDashboardNames::sLEFT_DRIVE_MOTOR_SPEED, mLeftMotorSpeed);
    SmartDashboard::PutNumber(SmartDashboardNames::sRIGHT_DRIVE_MOTOR_SPEED, mRightMotorSpeed);
}

void ASnobotDrivetrain::updateLog()
{
<<<<<<< HEAD
<<<<<<< HEAD
=======
>>>>>>> 408e130... Adding more CAN simulation
    mLogger->updateLogger(mLeftMotorDistance);
    mLogger->updateLogger(mRightMotorDistance);
    mLogger->updateLogger(mLeftMotorSpeed);
    mLogger->updateLogger(mRightMotorSpeed);
<<<<<<< HEAD
=======
//    mLogger.updateLogger(mLeftMotorDistance);
//    mLogger.updateLogger(mRightMotorDistance);
//    mLogger.updateLogger(mLeftMotorSpeed);
//    mLogger.updateLogger(mRightMotorSpeed);
>>>>>>> 9fafcd1... Adding a port of the robot code in CPP
=======
>>>>>>> 408e130... Adding more CAN simulation
}

double ASnobotDrivetrain::getRightDistance()
{
    return mRightMotorDistance;
}

double ASnobotDrivetrain::getLeftDistance()
{
    return mLeftMotorDistance;
}

void ASnobotDrivetrain::setLeftRightSpeed(double aLeftSpeed, double aRightSpeed)
{
    mLeftMotorSpeed = aLeftSpeed;
    mRightMotorSpeed = aRightSpeed;
    mRobotDrive.SetLeftRightMotorOutputs(mLeftMotorSpeed, mRightMotorSpeed);
}

double ASnobotDrivetrain::getLeftMotorSpeed()
{
    return mLeftMotorSpeed;
}

double ASnobotDrivetrain::getRightMotorSpeed()
{
    return mRightMotorSpeed;

}

void ASnobotDrivetrain::stop()
{
    setLeftRightSpeed(0, 0);
}

